src/internal/connector/exchange/container_resolver.go

```diff
@@ -40,50 +48,147 @@ type containersEnumerator interface {
 // folders if each folder is only a single character.
 const maxIterations = 300
 
-func newContainerResolver() *containerResolver {
+func newContainerResolver(refresher containerRefresher) *containerResolver {
 	return &containerResolver{
-		cache: map[string]graph.CachedContainer{},
+		cache:     map[string]graph.CachedContainer{},
+		refresher: refresher,
 	}
 }
 
 type containerResolver struct {
-	cache map[string]graph.CachedContainer
+	cache     map[string]graph.CachedContainer
+	refresher containerRefresher
 }
 
 func (cr *containerResolver) IDToPath(
 	ctx context.Context,
 	folderID string,
 ) (*path.Builder, *path.Builder, error) {
-	return cr.idToPath(ctx, folderID, 0)
+	ctx = clues.Add(ctx, "container_id", folderID)
+
+	c, ok := cr.cache[folderID]
+	if !ok {
+		return nil, nil, clues.New("folder not cached").WithClues(ctx)
+	}
+
+	p := c.Path()
+	if p == nil {
+		return nil, nil, clues.New("folder has no path").WithClues(ctx)
+	}
+
+	return p, c.Location(), nil
+}
+
+func (cr *containerResolver) refreshContainer(
+	ctx context.Context,
+	id string,
+) (graph.CachedContainer, bool, error) {
+	ctx = clues.Add(ctx, "refresh_container_id", id)
+	logger.Ctx(ctx).Debug("refreshing container")
+
+	if cr.refresher == nil {
+		return nil, false, clues.New("nil refresher")
+	}
+
+	c, err := cr.refresher.refreshContainer(ctx, id)
+	if err != nil && graph.IsErrDeletedInFlight(err) {
+		logger.Ctx(ctx).Debug("container deleted")
+		return nil, true, nil
+	} else if err != nil {
+		// This is some other error, just return it.
+		return nil, false, clues.Wrap(err, "refreshing container").WithClues(ctx)
+	}
+
+	return c, false, nil
 }
 
 func (cr *containerResolver) idToPath(
 	ctx context.Context,
 	folderID string,
 	depth int,
-) (*path.Builder, *path.Builder, error) {
+) (*path.Builder, *path.Builder, bool, bool, error) {
 	ctx = clues.Add(ctx, "container_id", folderID)
 
 	if depth >= maxIterations {
-		return nil, nil, clues.New("path contains cycle or is too tall").WithClues(ctx)
+		return nil, nil, false, false, clues.New("path contains cycle or is too tall").WithClues(ctx)
 	}
 
 	c, ok := cr.cache[folderID]
 	if !ok {
-		return nil, nil, clues.New("folder not cached").WithClues(ctx)
+		c, shouldDelete, err := cr.refreshContainer(ctx, folderID)
```

Comment:
This func (idToPath) is a lot to chew through.  Since this conditional block guarantees a return, it could help readability to move the internals to a new func and return a call to that func here.